template/base/.github/actions/setup/action.yml.njk

```diff
@@ -3,41 +3,127 @@ name: Setup environment
     required: true
+  rustfmt:
+    description: Install Rustfmt if `true`. Defaults to `false`.
+    required: false
   solana:
-    description: The Solana version to install
+    description: The Solana version to install. Skips if not provided.
+    required: false
 
 runs:
-  using: "composite"
+  using: 'composite'
   steps:
     - name: Setup pnpm
       uses: pnpm/action-setup@v3
+
     - name: Setup Node.js
       uses: actions/setup-node@v4
       with:
         node-version: {% raw %}${{ inputs.node }}{% endraw %}
-        cache: "pnpm"
-    - name: Install dependencies
+        cache: 'pnpm'
+
+    - name: Install Dependencies
       run: pnpm install --frozen-lockfile
       shell: bash
+
+    - name: Set Environment Variables
+      shell: bash
+      run: pnpm zx ./scripts/ci/set-env.mjs
+
+{% if solanaVersion === '2.0' %}
+    - name: Install Protobuf Compiler (Temporary Workaround for Solana 2.0)
+      if: {% raw %}${{ inputs.solana || inputs.rustfmt == 'true' || inputs.clippy == 'true' }}{% endraw %}
+      shell: bash
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y protobuf-compiler
+{% endif %}
+
+    - name: Install Rustfmt
+      if: {% raw %}${{ inputs.rustfmt == 'true' }}{% endraw %}
+      uses: dtolnay/rust-toolchain@master
+      with:
+        toolchain: {% raw %}${{ env.RUSTFMT_NIGHTLY_VERSION }}{% endraw %}
+        components: rustfmt
+
+    - name: Install Clippy
+      if: {% raw %}${{ inputs.clippy == 'true' }}{% endraw %}
+      uses: dtolnay/rust-toolchain@master
+      with:
+        toolchain: {% raw %}${{ env.CLIPPY_NIGHTLY_VERSION }}{% endraw %}
+        components: clippy
+
     - name: Install Solana
-      if: {% raw %}${{ inputs.solana != '' }}{% endraw %}
+      if: {% raw %}${{ inputs.solana }}{% endraw %}
       uses: metaplex-foundation/actions/install-solana@v1
       with:
         version: {% raw %}${{ inputs.solana }}{% endraw %}
-        cache: {% raw %}${{ inputs.cache }}{% endraw %}
+        cache: true
+
 {% if programFramework === 'anchor' %}
     - name: Install Anchor
       if: {% raw %}${{ inputs.anchor != '' }}{% endraw %}
       uses: metaplex-foundation/actions/install-anchor-cli@v1
       with:
         version: {% raw %}${{ inputs.anchor }}{% endraw %}
-        cache: {% raw %}${{ inputs.cache }}{% endraw %}
+        cache: true
 {% endif %}
+
+    - name: Cache Cargo Dependencies
+      if: {% raw %}${{ inputs.cargo-cache-key && !inputs.cargo-cache-fallback-key }}{% endraw %}
+      uses: actions/cache@v4
+      with:
+        path: |
+          ~/.cargo/bin/
+          ~/.cargo/registry/index/
+          ~/.cargo/registry/cache/
+          ~/.cargo/git/db/
+          target/
+        key: {% raw %}${{ runner.os }}-${{ inputs.cargo-cache-key }}-${{ hashFiles('**/Cargo.lock') }}{% endraw %}
+        restore-keys: {% raw %}${{ runner.os }}-${{ inputs.cargo-cache-key }}{% endraw %}
+
+    - name: Cache Cargo Dependencies With Fallback
```

Comment:
![dnr](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExNmw2ODJvMTFidzJ1c3JpeTNuOGxlZWczaWJkZDd5MmZuOG44Z203NiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/5xtDarGlbdGFzWZDT6o/giphy.gif)

Reply:
Haha yeah if you want to see that CI in action, I really recommend [checking this PR](https://github.com/solana-program/counter-shank/pull/1) instead. 🙂